orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,217 @@ common::Status HandleSharedInitializer(Graph& graph,
+  std::map<const NodeArg*, std::vector<NodeArg*>> tensor_replicas;
+  auto create_tensor_replica = [&tensor_replicas, &graph](const NodeArg* tensor,
+                                                          int consumer_stage) {
+    NodeArg& new_receive_output = CreateNodeArg(graph, *tensor);
+    const auto* old_shape = tensor->Shape();
+    if (old_shape != nullptr) {
+      new_receive_output.SetShape(*old_shape);
+    }
+    // Add value info for this newly added receive_output, for shape propagation
+    // when training this partition.
+    graph.AddValueInfo(&new_receive_output);
+    tensor_replicas[tensor][consumer_stage] = &new_receive_output;
+  };
+
+  // Checks whether the tensor is produced and consumed in the forward stage of
+  // the computation.
+  auto is_forward = [](const int producer_stage, const int consumer_stage) {
+    return producer_stage < consumer_stage;
+  };
+
+  // Checks whether the tensor is produced and consumed in the backward stage of
+  // the computation. 
+  auto is_backward = [](const int producer_stage, const int consumer_stage) {
+    return producer_stage > consumer_stage;
+  };
+
+  // Find tensors that need to be sent and forwarded.
+  for (const NodeArg* node_arg : initial_node_args) {
+    // Initialize tensor_replicas data structure.
+    auto inserted = tensor_replicas.emplace(
+                      std::make_pair(node_arg, std::vector<NodeArg*>(nstages)));
+    auto& replicas = (*inserted.first).second;
+
+    // TODO: for now we pretend that inputs are produced in stage 0,
+    // but I need to double check how they are handled.
+    int producer_stage = 0;
+    Node* producer_node = graph.GetMutableProducerNode(node_arg->Name());
+    assert(producer_node != nullptr);
+    producer_stage = op_to_stage.find(producer_node)->second;
+    
+    auto consumers = graph.GetMutableConsumerNodes(node_arg->Name());
+    if (consumers.size() == 0) {
+      continue;
+    }
+    
+    // This is only handling forwarding in the forward part of the graph.
+    int last_consumer_stage_fwd = -1;
+    for (Node* consumer : consumers) {
+      auto found_stage = op_to_stage.find(consumer);
+      assert(found_stage != op_to_stage.end());
+      int consumer_stage = found_stage->second;
+      // TODO: test case in which a tensor is produced by a fwd op, stashed and
+      // sent to the previous stage by a bwd op.
+      // For now, I'm assuming that if a tensor is produced by a fwd op and
+      // consumed by a bwd op, then producer and consumer are both in the same
+      // device. This will not always be the case though. 
+      if (!IsBackward(*producer_node) && IsBackward(*consumer)) {
+        // They must be on the same device. 
+        ORT_ENFORCE(producer_stage == consumer_stage,
+                    "Fwd producer and bwd consumer of a tensor must be in the same device.");
+      }
+
+      // It is impossible to have a bwd operator producing a tensor consumed
```

Comment:
>bwd [](start = 36, length = 3)

Please avoid abbreviation if possible. Let's replace "bwd" with "backward" in code and comments. Thank you.

Reply:
I believe I replaced all the fwd with forward and bwd with backward.